jaxrs-api/src/main/java/se/fortnox/reactivewizard/jaxrs/context/JaxRsRequestContext.java

```diff
@@ -0,0 +1,92 @@
+package se.fortnox.reactivewizard.jaxrs.context;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * A context that exists during handling of a jaxrs request.
+ *
+ * Any user code that wishes to interact with the context should do so using the {@link #setValue(Object, Object)} and
+ * {@link #getValue(Object)} methods.
+ */
+public class JaxRsRequestContext {
+    private static final Logger                           LOG     = LoggerFactory.getLogger(JaxRsRequestContext.class);
+    private static final ThreadLocal<JaxRsRequestContext> CONTEXT = new ThreadLocal<>();
+
+    private final Map<Object, Object> contextStorage;
+
+    private JaxRsRequestContext() {
+        contextStorage = new ConcurrentHashMap<>();
+    }
+
+    static JaxRsRequestContext getContext() {
+        return CONTEXT.get();
+    }
+
+    static void setContext(JaxRsRequestContext context) {
+        CONTEXT.set(context);
+    }
+
+    /**
+     * Retrieve a value stored in the context.
+     *
+     * @param key The key associated with the value
+     * @param <T> The value type
+     * @return An Optional containing the value. The Optional is empty if there is no value set,
+     *         if the value set is null or if the context is not open.
+     */
+    public static <T> Optional<T> getValue(Object key) {
+        JaxRsRequestContext ctx = getContext();
+        if (ctx == null) {
+            return Optional.empty();
+        }
+        return Optional.ofNullable((T) ctx.contextStorage.get(key));
+    }
+
+    /**
+     * Store a value in the context.
+     *
+     * @param key   The key associated with the value
+     * @param value The value
+     */
+    public static void setValue(Object key, Object value) {
+        JaxRsRequestContext ctx = getContext();
+        if (ctx == null) {
+            LOG.warn("No context exists or this is not executed in the context of a request");
+            return;
+        }
+        ctx.contextStorage.put(key, value);
+    }
+
+    /**
+     * Open a new context.
+     *
+     * The context does not support reentry and will throw an exception if a context already exists.
+     * User code should not need to open a context.
+     *
+     * @throws IllegalStateException if a context already exists
+     */
+    public static void open() {
+        if (CONTEXT.get() != null) {
+            throw new IllegalStateException("A context already exists");
```

Comment:
We should probably fail silently instead and log this. It would also be nice to have open and close as package private.

Reply:
I'm not sure. Since reentry is not supported and any failure would mean a programmatic error don't we want to fail spectacularly ASAP?

I thought about the visibility that but couldn't decide. It was easier having them public to make testing  easier. Will figure out a solution.

Reply:
So you decided to keep open/close public? I won't argue against, but just want to check that it wasn't forgotten.
Also, what's the reason for throwing exception when opening incorrectly, but not when closing?

Reply:
Yeah, I think it should probably be visible.

Opening twice may cause unexpected problems that are not immediately visible since it will clear the current context much like if it was closed (meaning, the properties are gone). Closing an already closed context does nothing and thus doesn't throw an exception.